plugins/parsers/xml/parser.go

```diff
@@ -0,0 +1,422 @@
+package xml
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/antchfx/xmlquery"
+	"github.com/antchfx/xpath"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+)
+
+type Parser struct {
+	Configs     []Config
+	DefaultTags map[string]string
+	Log         telegraf.Logger
+}
+
+type Config struct {
+	MetricName   string
+	MetricQuery  string            `toml:"metric_name"`
+	Selection    string            `toml:"metric_selection"`
+	Timestamp    string            `toml:"timestamp"`
+	TimestampFmt string            `toml:"timestamp_format"`
+	Tags         map[string]string `toml:"tags"`
+	Fields       map[string]string `toml:"fields"`
+	FieldsInt    map[string]string `toml:"fields_int"`
+
+	FieldSelection  string `toml:"field_selection"`
+	FieldNameQuery  string `toml:"field_name"`
+	FieldValueQuery string `toml:"field_value"`
+	FieldNameExpand bool   `toml:"field_name_expansion"`
+}
+
+func (p *Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	t := time.Now()
+
+	// Parse the XML
+	doc, err := xmlquery.Parse(strings.NewReader(string(buf)))
+	if err != nil {
+		return nil, err
+	}
+
+	// Queries
+	metrics := make([]telegraf.Metric, 0)
+	for _, config := range p.Configs {
+		if len(config.Selection) == 0 {
+			config.Selection = "/"
+		}
+		selectedNodes, err := xmlquery.QueryAll(doc, config.Selection)
+		if err != nil {
+			return nil, err
+		}
+		if len(selectedNodes) < 1 || selectedNodes[0] == nil {
+			p.debugEmptyQuery("metric selection", doc, config.Selection)
+			return nil, fmt.Errorf("cannot parse with empty selection node")
+		}
+
+		for _, selected := range selectedNodes {
+			m, err := p.parseQuery(t, doc, selected, config)
+			if err != nil {
+				return metrics, err
+			}
+
+			metrics = append(metrics, m)
+		}
+	}
+
+	return metrics, nil
+}
+
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	t := time.Now()
+
+	switch len(p.Configs) {
+	case 0:
+		return nil, nil
+	case 1:
+		config := p.Configs[0]
+
+		doc, err := xmlquery.Parse(strings.NewReader(line))
+		if err != nil {
+			return nil, err
+		}
+
+		selected := doc
+		if len(config.Selection) > 0 {
+			selectedNodes, err := xmlquery.QueryAll(doc, config.Selection)
+			if err != nil {
+				return nil, err
+			}
+			if len(selectedNodes) < 1 || selectedNodes[0] == nil {
+				p.debugEmptyQuery("metric selection", doc, config.Selection)
+				return nil, fmt.Errorf("cannot parse line with empty selection")
+			} else if len(selectedNodes) != 1 {
+				return nil, fmt.Errorf("cannot parse line with multiple selected nodes (%d)", len(selectedNodes))
+			}
+			selected = selectedNodes[0]
+		}
+
+		return p.parseQuery(t, doc, selected, config)
+	}
+	return nil, fmt.Errorf("cannot parse line with multiple (%d) configurations", len(p.Configs))
```

Comment:
@reimda What is the reason ParseLine cannot be used in combination with multiple configs?

Reply:
I'm not sure. It looks like p.document holds some state, so that p.document.Parse() can't be called for multiple configs. @srebhan wrote this code, maybe he can give a better answer.

Reply:
Oops, looks like you two are already discussing it in PR 11102